packages/runtime/src/controls/style.ts

```diff
@@ -28,7 +28,7 @@ type BorderStyle = 'dotted' | 'dashed' | 'solid'
 export type BorderSideShorthandPropertyData = {
   width: number | null | undefined
   style: BorderStyle
-  color?: ColorData | null
+  color?: ColorControlData | null
```

Comment:
So this is an interesting change. We want to be able to evolve the Style control independent of the Color control and while we were previous sharing the `ColorData` type, I'm not sure we should continue doing that. This `color` property doesn't come from the Color control, so it would be weird for it to have things specific to the Color control, like its data type.

I think that if this control was somehow using or composing the Color control that would be a different story. But in this case it wasn't, it was just sharing the same TypeScript type and I don't think it should have.

Something else to consider is that it seems `ColorData` was not defining the type for any specific control, like the Color control, but instead was a shared type for data used by multiple controls, like this one in addition to the Color control. We could continue to share `ColorData` and that could be used by both `ColorControlData` and this type and then certain utilities like `copyColorData` could be shared.

For example:

```ts
type ColorControlDataV0 = ColorData

type ColorControlDataV1 = { type: typeof ColorControlDataTypeV1; swatchId: string; alpha: number }

type ColorControlData = ColorControlDataV0 | ColorControlDataV1
```

```ts
color?: ColorData | null // this stays as `ColorData` instead of `ColorControlData`
```

This would result in no changes to the Style control while continuing to share the underlying color shape. Thoughts?

Reply:
I like but idea, but I want to see if we can _extend_ that thought a little further. If we intend to have some sort of base types, e.g. `color`, instead of using that base type directly in our interfaces, we extend that base type (create a new type) to use within the implementation?

e.g.

```ts
interface ColorData {
  swatchId: string;
  alpha: number;
}

// ....

interface ColorControlDataV0 extends ColorData {}

interface ColorControlDataV1 extends ColorData {
 type: typeof ColorControlDataTypeV1;
}
```

This way the types being consumed by the component can evolve more independently from the base type, while still sharing the underlying representation of it.

Reply:
@chanceaclark, I can’t seem to find a comment that I received an email for, but yeah, having interfaces and extending would work well. We could also use types and intersections if we’d rather use types over interfaces.

Reply:
I ended up removing the comment after reading your comment for the third time. It was pretty much close in nature. The only difference was types vs. interfaces, which when intersecting type assignments you get the same type rather than a new type. e.g.
![Screenshot 2024-01-02 at 15 24 11](https://github.com/makeswift/makeswift/assets/10539418/1f50535b-7005-4ddb-b3c7-3cc59563ace2)